File: go-api/pkg/matrix/validation.go
package matrix

// IsDiagonal checks if a matrix is diagonal
func IsDiagonal(matrix [][]int) bool {
	if len(matrix) == 0 {
		return false
	}

	rows := len(matrix)
	cols := len(matrix[0])

	if rows != cols {
		return false
	}

	for i := 0; i < rows; i++ {
		for j := 0; j < cols; j++ {
			if i != j && matrix[i][j] != 0 {
				return false
			}
		}
	}

	return true
}

// IsValid checks if matrix is valid (non-empty and rectangular)
func IsValid(matrix [][]int) bool {
	if len(matrix) == 0 {
		return false
	}

	cols := len(matrix[0])
	for i := 1; i < len(matrix); i++ {
		if len(matrix[i]) != cols {
			return false
		}
	}

	return true
}

File: node-api/src/utils/validation.js
const validation = {
  validateMatrix: (matrix) => {
    if (!Array.isArray(matrix)) {
      return { isValid: false, error: 'Input must be an array' };
    }

    if (matrix.length === 0) {
      return { isValid: false, error: 'Matrix cannot be empty' };
    }

    const firstRowLength = matrix[0].length;

    for (let i = 0; i < matrix.length; i++) {
      const row = matrix[i];

      if (!Array.isArray(row)) {
        return { isValid: false, error: `Row ${i} is not an array` };
      }

      if (row.length !== firstRowLength) {
        return { isValid: false, error: 'All rows must have the same length' };
      }

      // Validate that all elements are numbers
      for (let j = 0; j < row.length; j++) {
        if (typeof row[j] !== 'number' || isNaN(row[j])) {
          return { isValid: false, error: `Element at [${i}][${j}] is not a valid number` };
        }
      }
    }

    return { isValid: true };
  }
};

module.exports = validation;
